### ISSUE TITLE
View function can cause insufficient funds error

### ISSUE THREAD

**Issue by hellwolf** [trufflesuite/truffle#4457]
- [ ] I've [opened a support ticket](https://trufflesuite.zendesk.com/hc/en-us/requests/new) before filing this issue.

---------------------------

## Issue

In addition to #4397

We also noticed that sometimes view function can cause the following errors (on polygon, could be specific to their node?):

```
err: insufficient funds for gas * price + value: address 0xd66E40b0c30595bEc72153B502aC1E0c4785991B have 13763260682756217913 want 50000000000000000000000 (supplied gas 500000)
```

This occurs when using truffle, and set a default gas price in the network config, for example:

```
        matic: {
            gasPrice: +process.env.MATIC_GAS_PRICE || 30e9,
        },
```

The proposed solution is documented in https://github.com/trufflesuite/truffle/pull/4448

## Steps to Reproduce

https://github.com/hellwolf/truffle_contract_staticcall_gasprice_issue

        // This will trigger the bug, by using high default gasPrice
        // `GAS_PRICE=1e12 npx truffle --network matic exec scripts/test.js`
        // but without GAS_PRICE setting, it works
        // `npx truffle --network matic exec scripts/test.js`

```
$ npx truffle --network matic exec scripts/test.js 
Using network 'matic'.

usdcx name:  Super USDC (PoS)
```

```
$ GAS_PRICE=1e12 npx truffle --network matic exec scripts/test.js 
Using network 'matic'.

err: insufficient funds for gas * price + value: address 0x627306090abaB3A6e1400e9345bC60c78a8BEf57 have 62406659 want 600000000000000000000000000 (supplied gas 50000000)
Truffle v5.4.21 (core: 5.4.21)
Node v14.18.1
```

## Environment

* Operating System: Ubuntu 20
* Ethereum client: ?
* Truffle version (`truffle version`): v5.4.21
* node version (`node --version`): v14.18.1
* npm version (`npm --version`): ?

**Comment by haltman-at** [comment]
OK, I can reproduce the issue.  Thanks for bringing this up.  It definitely does seem that, in this instance, a call is failing with insufficient value.  Moreover, this is happening when using an external service, not Ganache, so this isn't a Ganache issue.  As such, it does look like there is something to be done here.

The approach taken in your PR #4397 is one possibility (although the PR itself will have to be fixed; as currently written it does nothing).  But I think we should make sure we understand the exact problem a bit better before taking any action.

Like -- why, exactly, is Matic refusing the transaction?  Is this truly a matter of ether in the sending account being insufficient, or is it a matter of the specified gas being insufficient?  Of course no amount of gas is explicitly being specified in this case, so it should be determined automatically via a gas estimate, but it something causing that to go wrong in this case?  (@fainashalts?)  There are a number of steps in the chain here (Truffle Test -> Truffle Contract -> Web3 -> the actual raw requests that are issued -> Matic), and it's not obvious to me which step is truly the problem.

So, while the approach you took in #4397 is (once fixed) certainly one possibility, I think we need to investigate and make sure we understand the problem a bit better before merging it.

**Comment by haltman-at** [comment]
Btw, one reason I'm wary of the approach that you took is, what if the user actually wants to set a `gasPrice` for a call?  Your solution prevents that.  This might not seem useful, but Truffle is often used for testing purposes, and people might expect this to work in their tests (which could check the gas price via the `GASPRICE` opcode and report what it finds).

I mean, if this truly ends up being the best place to fix it then we may just have to live with that, but I want to be sure it *is* the appropriate place first!

**Comment by hellwolf** [comment]
> Btw, one reason I'm wary of the approach that you took is, what if the user actually wants to set a `gasPrice` for a call? Your solution prevents that. This might not seem useful, but Truffle is often used for testing purposes, and people might expect this to work in their tests (which could check the gas price via the `GASPRICE` opcode and report what it finds).
> 
> I mean, if this truly ends up being the best place to fix it then we may just have to live with that, but I want to be sure it _is_ the appropriate place first!

Sorry, my typo. Fixed it.

It should work in general, because of the side effect of `skipNetworkCheck`, which actually the pull request #4397 original intention was to indicate that `prepareCall` was called by `call`, and in that case, some extra network IOs should be avoided. That's why in case of `send`, the gas price can still be set by the user.

**Comment by haltman-at** [comment]
OK, I've started taking more of a look at this -- sorry, been busy.  I can confirm that the issue is not confined to Matic; the same thing happens when I try it on Goerli.  Will say more once I've looked more.

**Comment by haltman-at** [comment]
Also, the problem doesn't only occur with `gasPrice`; it also occurs with `maxFeePerGas` and `maxPriorityFeePerGas`.

**Comment by haltman-at** [comment]
OK, I think I see what's going on here.  I'm going to go properly review the PR now that I have a better understanding of what's going on!

**Comment by haltman-at** [comment]
I should also note that the issue ocurs even when using `web3` directly (assuming we manually set the parameters).  So this isn't an issue in HDWalletProvider or anything like that.  You were right to open the PR against Truffle Contract, although (as I discuss there) I think what you're doing actually belongs in a different location within Truffle Contract.

**Comment by haltman-at** [comment]
This should now be fixed in 5.4.29, but let us know if there's still a problem!
